File: android/ui/src/main/java/ru/maksimbulva/ui/chessboard/items/ChessboardAdapter.kt
package ru.maksimbulva.ui.chessboard.items

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import ru.maksimbulva.ui.chessboard.ChessboardSquareView
import ru.maksimbulva.ui.chessboard.ChessboardTheme

internal class ChessboardAdapter(
    private val context: Context,
    private val theme: ChessboardTheme
) : BaseAdapter() {

    private var items: List<ChessboardItem> = emptyList()

    override fun getCount() = items.size

    override fun getItemId(index: Int) = index.toLong()

    override fun getItem(index: Int) = items.getOrNull(index)

    override fun getView(index: Int, convertView: View?, parent: ViewGroup?): View {
        val itemView = if (convertView != null && convertView is ChessboardSquareView) {
            convertView
        } else {
            ChessboardSquareView(context)
        }

        return itemView.also {
            it.setTheme(theme)
            it.bind(items[index])
        }
    }

    fun setItems(items: List<ChessboardItem>) {
        this.items = items.toList()
        notifyDataSetChanged()
    }
}

File: ChessApp/UI/Internal/Chessboard/ChessboardAdapter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using UI.Items;

namespace UI.Internal.Chessboard
{
    internal class ChessboardAdapter
        : BaseAdapter
    {
        private readonly Context _context;

        private readonly List<ChessboardItem> items = new List<ChessboardItem>();

        public ChessboardAdapter(Context context)
        {
            _context = context;
        }

        public override int Count => items.Count;

        public override Java.Lang.Object GetItem(int position) => items[position];

        public override long GetItemId(int position) => 0;

        public override View GetView(int position, View convertView, ViewGroup parent)
        {
            ChessboardSquareView view = convertView as ChessboardSquareView;
            if (view == null)
            {
                view = new ChessboardSquareView(_context, null);
            }
            view.Bind(items[position]);
            return view;
        }

        public void SetItems(IEnumerable<ChessboardItem> items)
        {
            this.items.Clear();
            this.items.AddRange(items);
            NotifyDataSetChanged();
        }
    }
}
